# Data Cleaning with Pandas

In [187]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [188]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/Users/jjherranzsarrion/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [189]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [190]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [191]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 62.0+ MB


In [192]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Users/jjherranzsarrion/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [193]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [194]:
bldg_df.head().T

0                          1  \
Major                                   14800                      14800   
Minor                                     168                        425   
BldgNbr                                     1                          1   
NbrLivingUnits                              1                          1   
Address             3047   63RD AVE SW  98116  3231   63RD AVE SW  98116   
BuildingNumber                          3047                       3231    
Fraction                                                                   
DirectionPrefix                                                            
StreetName          63RD                       63RD                        
StreetType                             AVE                        AVE      
DirectionSuffix                            SW                         SW   
ZipCode                                 98116                      98116   
Stories                                     1                          2   
BldgGrade                                   7                          8   
BldgGradeVar                                0                          0   
SqFt1stFloor                              750                       1010   
SqFtHalfFloor                               0                          0   
SqFt2ndFloor                                0                        560   
SqFtUpperFloor                              0                          0   
SqFtUnfinFull                               0                          0   
SqFtUnfinHalf                               0                          0   
SqFtTotLiving                            1500                       2070   
SqFtTotBasement                           750                        680   
SqFtFinBasement                           750                        500   
FinBasementGrade                            7                          7   
SqFtGarageBasement                          0                          0   
SqFtGarageAttached                          0                          0   
DaylightBasement                            N                              
SqFtOpenPorch                              20                         60   
SqFtEnclosedPorch                           0                          0   
SqFtDeck                                    0                        290   
HeatSystem                                  5                          5   
HeatSource                                  2                          2   
BrickStone                                  0                          0   
ViewUtilization                             N                              
Bedrooms                                    3                          4   
BathHalfCount                               0                          1   
Bath3qtrCount                               0                          1   
BathFullCount                               2                          1   
FpSingleStory                               0                          0   
FpMultiStory                                0                          1   
FpFreestanding                              0                          1   
FpAdditional                                0                          0   
YrBuilt                                  1949                       1924   
YrRenovated                                 0                          0   
PcntComplete                                0                          0   
Obsolescence                                0                          0   
PcntNetCondition                            0                          0   
Condition                                   4                          5   
AddnlCost                                3000                          0   

                                            2                          3  \
Major                                   14800                      14800   
Minor          

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [195]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [196]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512387 entries, 0 to 512386
Data columns (total 4 columns):
Major            512387 non-null int64
Minor            512387 non-null int64
SqFtTotLiving    512387 non-null int64
ZipCode          468443 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [233]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [198]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 943879

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [199]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [200]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [201]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 62.0+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [202]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [203]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 62.0+ MB


Now, let's try our join again.

In [283]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [249]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [284]:
import numpy as np
nonzero_mask = (sales_data['SalePrice'] > 0)
np.percentile(sales_data['SalePrice'][nonzero_mask], 4)

65000.0

In [250]:
sales_mini = sales_data.iloc[-1000:].copy()

In [251]:
sales_mini['Date'] = sales_mini['DocumentDate'].apply(pd.Timestamp)

In [252]:
sales_mini.head().T

1449456              1449457              1449458  \
Major                       247348               742170               352890   
Minor                         1050                   50                   87   
DocumentDate            05/13/1985           06/11/1992           07/17/1987   
SalePrice                    30000               197000                94500   
SqFtTotLiving                 2890                 2310                 1240   
ZipCode                      98058                98033                98109   
Date           1985-05-13 00:00:00  1992-06-11 00:00:00  1987-07-17 00:00:00   

                           1449459              1449460  
Major                        62106               551560  
Minor                         9021                  142  
DocumentDate            06/30/2016           06/02/2018  
SalePrice                   286000               325000  
SqFtTotLiving                  960                 1380  
ZipCode                      98042                98001  
Date           2016-06-30 00:00:00  2018-06-02 00:00:00

In [253]:
ts = pd.Timestamp('05/13/1985')

In [254]:
ts.year

1985

In [255]:
sales_mini['Year'] = sales_mini['Date'].apply(lambda ts: ts.year)

In [256]:
def get_year_from_text_date(text_date):
    ts = pd.Timestamp(text_date)
    return ts.year

In [257]:
sales_mini['Year'] = sales_mini['Date'].apply(get_year_from_text_date)

In [258]:
sales_mini.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 1449456 to 1450455
Data columns (total 8 columns):
Major            1000 non-null float64
Minor            1000 non-null float64
DocumentDate     1000 non-null object
SalePrice        1000 non-null int64
SqFtTotLiving    1000 non-null int64
ZipCode          895 non-null object
Date             1000 non-null datetime64[ns]
Year             1000 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 70.3+ KB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [259]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91   717370.0   350.0   12/01/1997          0           3380     NaN
92   717370.0   350.0   09/13/2004     300000           3380     NaN
93   717370.0   350.0   02/06/2006     901000           3380     NaN
110  277110.0  1923.0   02/08/2007     372500           1000     NaN
111  277110.0  1923.0   02/08/2007          0           1000     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [285]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [286]:
sales_data = sales_data.loc[~(sales_data['SalePrice'] == 0), :]
sales_data

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0        138860.0   110.0   08/21/2014     245000           1490   98002
1        138860.0   110.0   06/12/1989     109300           1490   98002
2        138860.0   110.0   01/16/2005      14684           1490   98002
6        423943.0    50.0   07/15/1999      96000            960   98092
7        423943.0    50.0   01/08/2001     127500            960   98092
11       403700.0   715.0   07/03/2013     464500           1780   98008
12       403700.0   715.0   02/21/2013     357000           1780   98008
13       403700.0   715.0   10/13/1995     142000           1780   98008
14       403700.0   715.0   02/22/2007     528000           1780   98008
16        98400.0   380.0   03/23/1994     139950           2050   98058
17        98400.0   380.0   03/29/2017     439500           2050   98058
19       330395.0  1610.0   06/07/2012     335000           2210   98038
20       330395.0  1610.0   08/26/2013     369900           2210   98038
21       330395.0  1610.0   06/27/1994     192950           2210   98038
22       330395.0  1610.0   08/22/2013     369900           2210   98038
23       330395.0  1610.0   11/10/1993     107500           2210   98038
25       675260.0   940.0   03/06/1998     209950           2040   98031
26       675260.0   940.0   08/01/1995     191950           2040   98031
27       675260.0   940.0   09/13/2004     289950           2040   98031
29       795508.0   170.0   04/09/2001     174000            980   98058
31       795508.0   170.0   10/31/1996     135000            980   98058
32       795508.0   170.0   04/30/2004     210000            980   98058
33       795508.0   170.0   12/21/2012     148000            980   98058
34       638580.0   110.0   07/14/2015     190000           2130   98188
36       638580.0   110.0   12/21/2015     338000           2130   98188
38       252206.0  9011.0   02/21/1997     300000           2530   98051
39       252206.0  9011.0   11/28/2000     390000           2530   98051
40       321165.0    50.0   05/10/2004    2100000           1800   98001
42       321165.0    50.0   10/24/2013     215000           1800   98001
44       321165.0    50.0   09/20/2004     270000           1800   98001
...           ...     ...          ...        ...            ...     ...
1450412  701220.0    85.0   09/23/2002     350000           1040   98119
1450413  736960.0    45.0   03/25/2011     535000           2620   98199
1450414  124500.0  3445.0   08/03/2017     877000           1100   98033
1450415  795030.0    10.0   03/23/1999     450000           2270   98118
1450417  872210.0   380.0   04/22/1992     290000           3120   98112
1450418  674670.0   305.0   03/14/2000     172000           1760   98105
1450419  773230.0   260.0   09/26/1984     147209           2880   98052
1450421  132832.0   640.0   03/01/1984     109950           2480   98058
1450422  247100.0    70.0   07/25/1990     350000           4950   98028
1450423  326230.0   545.0   01/13/1993     245000           2270   98039
1450425  212303.0  9015.0   09/25/1989     145000           1550   98070
1450427  882590.0   650.0   11/12/2003     412300           2720   98115
1450428  322480.0   155.0   03/16/1999     138000           1260   98002
1450429  387630.0   880.0   12/09/1997     224000           2280   98034
1450430  531610.0   390.0   01/22/1993     200000           2630   98112
1450431  717730.0   740.0   06/12/1985      96000           1460   98115
1450434  639200.0  2804.0   05/03/2007     696000           2480   98115
1450435  885670.0   200.0   02/19/1982     136373           3180   98052
1450436   19230.0  1100.0   07/21/1999     277000           1790   98068
1450437  440600.0   130.0   10/27/1983      62500            960   98058
1450439  892410.0    30.0   03/05/1996     357000           2960   98115
1450440  778536.0   250.0   09/15/1987     335000           3250   98177
1450441  287210.0  1240.0   12/09/2016     535000           1170   981

In [287]:
sales_data = sales_data.loc[~(sales_data['SqFtTotLiving'] == 0), :]
sales_data

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0        138860.0   110.0   08/21/2014     245000           1490   98002
1        138860.0   110.0   06/12/1989     109300           1490   98002
2        138860.0   110.0   01/16/2005      14684           1490   98002
6        423943.0    50.0   07/15/1999      96000            960   98092
7        423943.0    50.0   01/08/2001     127500            960   98092
11       403700.0   715.0   07/03/2013     464500           1780   98008
12       403700.0   715.0   02/21/2013     357000           1780   98008
13       403700.0   715.0   10/13/1995     142000           1780   98008
14       403700.0   715.0   02/22/2007     528000           1780   98008
16        98400.0   380.0   03/23/1994     139950           2050   98058
17        98400.0   380.0   03/29/2017     439500           2050   98058
19       330395.0  1610.0   06/07/2012     335000           2210   98038
20       330395.0  1610.0   08/26/2013     369900           2210   98038
21       330395.0  1610.0   06/27/1994     192950           2210   98038
22       330395.0  1610.0   08/22/2013     369900           2210   98038
23       330395.0  1610.0   11/10/1993     107500           2210   98038
25       675260.0   940.0   03/06/1998     209950           2040   98031
26       675260.0   940.0   08/01/1995     191950           2040   98031
27       675260.0   940.0   09/13/2004     289950           2040   98031
29       795508.0   170.0   04/09/2001     174000            980   98058
31       795508.0   170.0   10/31/1996     135000            980   98058
32       795508.0   170.0   04/30/2004     210000            980   98058
33       795508.0   170.0   12/21/2012     148000            980   98058
34       638580.0   110.0   07/14/2015     190000           2130   98188
36       638580.0   110.0   12/21/2015     338000           2130   98188
38       252206.0  9011.0   02/21/1997     300000           2530   98051
39       252206.0  9011.0   11/28/2000     390000           2530   98051
40       321165.0    50.0   05/10/2004    2100000           1800   98001
42       321165.0    50.0   10/24/2013     215000           1800   98001
44       321165.0    50.0   09/20/2004     270000           1800   98001
...           ...     ...          ...        ...            ...     ...
1450412  701220.0    85.0   09/23/2002     350000           1040   98119
1450413  736960.0    45.0   03/25/2011     535000           2620   98199
1450414  124500.0  3445.0   08/03/2017     877000           1100   98033
1450415  795030.0    10.0   03/23/1999     450000           2270   98118
1450417  872210.0   380.0   04/22/1992     290000           3120   98112
1450418  674670.0   305.0   03/14/2000     172000           1760   98105
1450419  773230.0   260.0   09/26/1984     147209           2880   98052
1450421  132832.0   640.0   03/01/1984     109950           2480   98058
1450422  247100.0    70.0   07/25/1990     350000           4950   98028
1450423  326230.0   545.0   01/13/1993     245000           2270   98039
1450425  212303.0  9015.0   09/25/1989     145000           1550   98070
1450427  882590.0   650.0   11/12/2003     412300           2720   98115
1450428  322480.0   155.0   03/16/1999     138000           1260   98002
1450429  387630.0   880.0   12/09/1997     224000           2280   98034
1450430  531610.0   390.0   01/22/1993     200000           2630   98112
1450431  717730.0   740.0   06/12/1985      96000           1460   98115
1450434  639200.0  2804.0   05/03/2007     696000           2480   98115
1450435  885670.0   200.0   02/19/1982     136373           3180   98052
1450436   19230.0  1100.0   07/21/1999     277000           1790   98068
1450437  440600.0   130.0   10/27/1983      62500            960   98058
1450439  892410.0    30.0   03/05/1996     357000           2960   98115
1450440  778536.0   250.0   09/15/1987     335000           3250   98177
1450441  287210.0  1240.0   12/09/2016     535000           1170   981

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [303]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def clean_zips(x):
    x = str(x)
    if len(x) >= 5:
        x = x[:5]
        return round(int(x),0)
    else:
        return np.nan

sales_data_copy = sales_data.copy()

sales_data_copy['Temp'] = sales_data_copy['ZipCode'].apply(clean_zips)

sales_data_copy['Temp'].isna().sum()

21

In [ ]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_data['ZipCode'] = sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False]

zip_column = sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode']
zip_column.head()

zip_column = zip_column.apply(lambda zips: zips[:5] if len(zips) > 5 else np.nan)

In [289]:
#sales_data = sales_data.loc[sales_data['ZipCode'].apply(str)]
zip_column.apply(str)

17669      98042-3001
17670      98042-3001
17671      98042-3001
17672      98042-3001
21350      98199-3014
21351      98199-3014
21352      98199-3014
31310      98033-4917
31311      98033-4917
31312      98033-4917
31313      98033-4917
31314      98033-4917
58775      98031-3173
71473      98042-3001
71474      98042-3001
75157      98136-1728
75158      98136-1728
75159      98136-1728
75160      98136-1728
111115     98052-1963
120874     98074-6315
120875     98074-6315
120879     98074-6315
130253     98058-9018
130254     98058-9018
152163     98177-4810
152165     98177-4810
193979     98023-2712
193981     98023-2712
201372     98074-3438
              ...    
1197925    98042-3001
1217235    98074-4092
1263738    98146-3008
1263739    98146-3008
1269239    98059-7120
1270414    98042-3001
1282477    98042-3001
1282478    98042-3001
1311008            WA
1311009            WA
1326809    98042-8215
1326810    98042-8215
1326811    98042-8215
1326812    98042-8215
1332675   

In [291]:
zip_column = zip_column.apply(lambda zips: zips[:5] if len(zips) > 5 else np.nan)

In [295]:
zip_column = zip_column.loc[~zip_column.isna()]
zip_column

17669      98042
17670      98042
17671      98042
17672      98042
21350      98199
21351      98199
21352      98199
31310      98033
31311      98033
31312      98033
31313      98033
31314      98033
58775      98031
71473      98042
71474      98042
75157      98136
75158      98136
75159      98136
75160      98136
111115     98052
120874     98074
120875     98074
120879     98074
130253     98058
130254     98058
152163     98177
152165     98177
193979     98023
193981     98023
201372     98074
           ...  
1191511    98058
1197923    98042
1197924    98042
1197925    98042
1217235    98074
1263738    98146
1263739    98146
1269239    98059
1270414    98042
1282477    98042
1282478    98042
1326809    98042
1326810    98042
1326811    98042
1326812    98042
1332675    98028
1332676    98028
1332677    98028
1338309    98028
1347061    98075
1363973    98042
1363974    98042
1387612    98042
1400334    98023
1403176    98045
1413305    98074
1418660    98028
1420222    980

### 3. Add a column for PricePerSqFt



### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!